corehq/apps/app_manager/tests/test_schedule.py

```diff
@@ -0,0 +1,408 @@
+        phase.remove_form(form_1)
+
+        self.assertEqual(len(phase.forms), 1)
+        self.assertEqual([form_2], list(phase.get_forms()))
+
+        with self.assertRaises(ScheduleError):
+            phase.remove_form(form_1)
+
+        # Removing all the forms deletes the phase
+        phase.remove_form(form_2)
+        self.assertEqual(len(self.module.schedule_phases), 0)
+
+    def test_add_form_to_phase(self):
+        self.module.schedule_phases = [
+            SchedulePhase(
+                anchor='dob',
+                forms=[SchedulePhaseForm(form_id=self.form_1.unique_id),
+                       SchedulePhaseForm(form_id=self.form_2.unique_id)]
+            ),
+            SchedulePhase(anchor='second_phase', forms=[]),
+        ]
+        phases = list(self.module.get_schedule_phases())
+        phase1 = phases[0]
+        phase1.add_form(self.form_3)
+        self.assertEqual(phase1.get_phase_form_index(self.form_3), 2)
+
+        # adding a form to a different phase removes it from the first phase
+        phase2 = phases[1]
+        phase2.add_form(self.form_3)
+        self.assertEqual(phase2.get_phase_form_index(self.form_3), 0)
+        self.assertIsNone(phase1.get_form(self.form_3))
+
+    def test_schedule_detail(self):
+        self._apply_schedule_phases()
+
+        suite = self.app.create_suite()
+        self.assertXmlPartialEqual(self.get_xml('schedule-entry'), suite, "./detail[@id='m1_case_short']")
+
+    def test_schedule_fixture(self):
+        self._apply_schedule_phases()
+
+        suite = self.app.create_suite()
+        self.assertXmlPartialEqual(self.get_xml('schedule-fixture'), suite, './fixture')
+
+    def test_multiple_modules(self):
+        self._apply_schedule_phases()
+
+        other_module = self.app.get_module(2)
+        other_module.has_schedule = True
+        scheduled_form = other_module.get_form(0)
+        scheduled_form.schedule = FormSchedule(
+            visits=[
+                ScheduleVisit(due=9),
+                ScheduleVisit(due=11)
+            ]
+        )
+        other_module.forms.append(copy.copy(scheduled_form))
+
+        other_module.schedule_phases = [
+            SchedulePhase(
+                anchor='case_property',
+                forms=[SchedulePhaseForm(form_id=scheduled_form.unique_id)]
+            )
+        ]
+
+        expected_fixture = """
+             <partial>
+             <fixture id="schedule:m2:p1:f0">
+                 <schedule expires="" allow_unscheduled="False">
+                     <visit id="1" due="9" repeats="False"/>
+                     <visit id="2" due="11" repeats="False"/>
+                 </schedule>
+             </fixture>
+             </partial>
+        """
+
+        suite = self.app.create_suite()
+
+        self.assertXmlPartialEqual(expected_fixture, suite, './fixture[@id="schedule:m2:p1:f0"]')
+        self.assertXmlHasXpath(suite, './fixture[@id="schedule:m1:p1:f0"]')
+
+    def test_form_filtering(self):
```

Comment:
Other reviewers: this is a useful test to read to understand what this is doing.